Twitter風掲示板に「RT」と「いいね！」機能を実装

U-11 opened the pull request:
index.php に機能実装
view.phpもRTの内容が表示されるように変更

U-11 committed:
「いいね！」機能の実装
ボタン押下を繰り返すと、いいね数が+1、-1される

U-11 committed:
「リツイート」ボタンを押下するとRTすることができる
ボタン押下を繰り返すと、RTの取り消しを行うことができる
RTされた件数が分かる
「いいね！」機能の不具合の習性

U-11 committed:
vew.phpはRTも表示出来るように修正
その他のPHPファイルは教科書通り実装

U-11 committed:
インデントの調整等

gacha1987 commented on php-rt-iine/delete.php:
```diff
@@ -0,0 +1,19 @@
+if(isset($_SESSION['id'])){
+
+  $messages=$db->prepare('SELECT * FROM posts WHERE id=?');
+  $messages->execute(array($_REQUEST['id']));
```
`$_REQUEST` の値をそのまま使用しているため、
ユーザーが数値以外や悪意ある入力をした場合もSQLが実行され、
エラーやセキュリティ事故を起こしてしまいますので、
事前にクエリパラメータが数値かやDBに値がそもそも存在するかチェックする処理を追加しましょう
※バリデーション と言います

ユーザーからパラメータ受ける箇所は必ずチェックするようにしてください

gacha1987 commented on php-rt-iine/index.php:
```diff
@@ -0,0 +1,328 @@
+// 通常の投稿の登録
+if(!empty($_POST['message'])){
+  if(empty($_POST['rt_post_id'])){
+    if($_POST['message']!=''){
```
比較をする際は `===`や`!==`などデータ型まで含めた判定を行いましょう。
曖昧な比較にすると意図せぬ箇所でバグが発生する場合があります。

また、`$_POST['message']!=''`という判定はL25の`!empty($_POST['message'])`という判定が真の時点で空文字は入ってこないので不要と思います。
https://www.php.net/manual/ja/types.comparisons.php

gacha1987 commented on php-rt-iine/index.php:
```diff
@@ -0,0 +1,328 @@
+    }
+  }else{
+    // コメントありリツイートをDBに登録
+    if($_POST['message']!=''){
```
こちらもL25の!empty($_POST['message'])という判定が真の時点で空文字は入ってこないので不要と思います。

gacha1987 commented on php-rt-iine/index.php:
```diff
@@ -0,0 +1,328 @@
+if(!empty($_POST['message'])){
+  if(empty($_POST['rt_post_id'])){
+    if($_POST['message']!=''){
+      $message=$db->prepare('INSERT INTO posts SET message=?,member_id=?,created=NOW()');
```
頂いたSQL情報だと `is_deleted ` はDEFAULT値の指定がないため、NULLになってしまうのでこの処理だと画面上に表示されません。
SQL情報を修正して展開いただくか、INSERT文を修正してください。

DDL情報を渡す場合は、それを実行するだけで正しく動くようなDDL文を書くのが理想です。

gacha1987 commented on php-rt-iine/index.php:
```diff
@@ -0,0 +1,328 @@
+      $com_rt->execute();
+
+    // リツイートカウント
+      $com_rt_cnt=$db->prepare('UPDATE posts SET rt_count=rt_count+1 WHERE id=?');
```
頂いたDDLですと`rt_count `はNULLのままで正しくカウントアップされないので、頂いたDDLを修正するか処理を見直してください。
like_countも同様です。

gacha1987 commented on php-rt-iine/join/thanks.php:
```diff
@@ -0,0 +1,26 @@
+    <p>ユーザー登録が完了しました。</p>
+    <p><a href="../login.php">ログインする</a></p>
+
+    </form>
```
閉じタグのみあるので削除してください

gacha1987 reviewed (commented):
コメントしました。
頂いたDDLでテーブルを作って確認すると正しく動かない箇所が多いため、提供したDDLとUさんの方で使っているテーブルに情報差異がないか確認し、DDLが違っていた場合は再度レビュー依頼時に提供してください。
(DEFAULTとかがないためエラーになっている箇所が多いので恐らくそこら辺の定義が抜けているかと)

また、上記の影響で出ているかもしれませんが、NOTICEエラーが出ているので確認して必要あれば修正してください。
```
[18-Oct-2020 07:27:57 Europe/Berlin] PHP Notice:  Undefined variable: message in /Applications/MAMP/htdocs/mizutaniyu_php/php-rt-iine/index.php on line 201
[18-Oct-2020 07:27:57 Europe/Berlin] PHP Notice:  Undefined index: res in /Applications/MAMP/htdocs/mizutaniyu_php/php-rt-iine/index.php on line 202
[18-Oct-2020 07:27:57 Europe/Berlin] PHP Notice:  Undefined index: retweet in /Applications/MAMP/htdocs/mizutaniyu_php/php-rt-iine/index.php on line 222
```

U-11 committed:
Noticeエラー箇所のNULL判定の見直し

U-11 committed:
バリデーションの実装
閉じタグの削除

U-11 committed:
不要なタグの削除

U-11 committed:
厳密な演算子に修正
メールアドレスにバリデーションを実装

U-11 committed:
厳密な比較演算子に修正

U-11 committed:
閉じタグの削除

U-11 committed:
バリデーションの実装

U-11 committed:
DB挿入前にバリデーションを実装
不要なif文の削除

U-11 commented:
以下、DDL全文です。

CREATE TABLE members
 (id INT AUTO_INCREMENT PRIMARY KEY,
name VARCHAR(255),
email VARCHAR(255),
password VARCHAR(100),
picture VARCHAR(255),  
created DATETIME,
modified timestamp);

CREATE TABLE posts
 (id INT AUTO_INCREMENT PRIMARY KEY,
message TEXT,
member_id INT,
reply_post_id INT,
rt_count INT NOT NULL DEFAULT '0',
like_count INT NOT NULL DEFAULT '0',
created DATETIME,
modified timestamp,
is_deleted INT NOT NULL DEFAULT '0');

CREATE TABLE retweets
(id INT AUTO_INCREMENT PRIMARY KEY,  
rt_member_id INT,  
rt_id INT,
post_id INT,
post_member_id INT,
post TEXT,  
created DATETIME,  
modified TIMESTAMP,
is_deleted INT NOT NULL DEFAULT '0');

CREATE TABLE likes
(id INT AUTO_INCREMENT PRIMARY KEY,  
like_member_id INT,  
post_id INT,
created DATETIME,  
modified TIMESTAMP,
is_deleted INT NOT NULL DEFAULT '0');

gacha1987 commented on php-rt-iine/index.php:
```diff
@@ -0,0 +1,339 @@
+if($page===''){
+  $page=1;
+}
+$page=max($page,1);
```
L59〜L65は1行にまとめられると思うので、ちょっと考えてみてください。ヒント：三項演算子
ここの処理でやりたいことは下記だと思うのでこれを式に当て嵌めてみてください。
 `クエリパラメータが存在し、かつ数値だったらクエリパラメータを変数$pageに代入、そうでない場合は1(デフォルト値)を変数$pageに代入`

U-11 replied:
三項演算子を使って条件を1行にまとめました。

gacha1987 commented on php-rt-iine/index.php:
```diff
@@ -0,0 +1,339 @@
+              <a href="view.php?id=<?php echo h($post['reply_post_id']); ?>">返信元のメッセージ</a>
+              <?php endif; ?>
+              <?php if($_SESSION['id']===$post['member_id']): ?>
+              <a href="delete.php?id=<?php echo h($post['id']); ?>" style="color:#f33;">削除</a>
```
リツイート投稿をここの削除ボタンで消した場合にリツイート数が減らないので、リツイート投稿を消す場合は元のリツイート数を減算する考慮を追加してください。

U-11 replied:
リツイート投稿を削除ボタンで消した場合も元のリツイート数を減算するようdelete.phpファイルで実装しました。

gacha1987 commented on php-rt-iine/join/index.php:
```diff
@@ -0,0 +1,126 @@
+    }
+    // 画像に命名
+    $image=date('YmdHis').$_FILES['image']['name'];
+    move_uploaded_file($_FILES['image']['tmp_name'],'../member_picture/'.$image);
```
member_pictureディレクトリがなく、画像が保存できないのでgitのコミット対象に含めるようにしてください。

U-11 replied:
member_pictureディレクトリを追加しました。

gacha1987 commented on php-rt-iine/index.php:
```diff
@@ -0,0 +1,339 @@
+    exit();
+  }else{
+    if(is_numeric($_POST['rt_post_id']) && is_numeric($_POST['post_member_id'])){
+      // コメントありリツイートをDBに登録
```
リツイート処理やいいね処理は`delete.php `みたいに処理を別ファイルに分けましょう。
それをindex.phpから呼ぶようにしてあげた方がindex.phpが肥大化せず読みやすいコードになります

U-11 replied:
rt_action.phpとlike_action.phpに分けて実装し直しました。

gacha1987 reviewed (commented):
コメントしました。

U-11 committed:
L59〜L65の条件を三項演算子で1行にまとめて実装

U-11 committed:
member_pictureフォルダの追加

U-11 committed:
リツイートの投稿を削除した場合に、元の投稿のリツイート数を減算するよう実装

U-11 committed:
いいね処理をlike_action.phpで実装

U-11 committed:
リツイート処理をrt_action.phpで実装

U-11 committed:
Delete 20200927153714848369.jpg

U-11 committed:
リツイート処理をrt_action.phpで実装

U-11 committed:
Merge branch 'feature/rt-iine' of https://github.com/U-11/mizutaniyu_php into feature/rt-iine

U-11 committed:
Create .gitkeep

gacha1987 commented on php-rt-iine/index.php:
```diff
@@ -0,0 +1,220 @@
+}
+
+// ページング(１ページ５投稿)
+isset($_REQUEST['page']) && is_numeric($_REQUEST['page']) ? $page=$_REQUEST['page'] : $page=max($page,1);
```
これでも動きますが、三項演算子の書き方として`$page = a ? b : c;`という風に、式としては値のみ返して1つの変数に代入を行う方が可読性が上がり、保守しやすくなります。
多項演算子は様々な書き方ができますが、極力シンプルにするとgoodです。
https://www.php.net/manual/ja/language.operators.comparison.php#language.operators.comparison.ternary

また、`$page=max($page,1);`の引数になっている$pageはそもそも存在していない変数でNOTICEエラーが出ているため修正してください。
→数値でない場合のルートなのでmax変数使わずにデフォルト値をそのまま入れてあげればいいかなと思います。

U-11 replied:
三項演算子を可読性の高い書き方に修正いたしました。

gacha1987 commented on php-rt-iine/index.php:
```diff
@@ -0,0 +1,220 @@
+          <dl>
+            <dt><?php echo h($member['name']); ?>さん、メッセージをどうぞ</dt>
+            <dd><textarea name="message" cols="50" rows="5" <?php if(isset($_POST['retweet'])){echo 'placeholder="コメントをつけてリツイート"';} ?>><?php echo h($message) ?? NULL; ?></textarea>
+            <input type="hidden" name="reply_post_id" value="<?php echo h($_REQUEST['res']) ?? NULL; ?>"></dd>
```
下記のNOTICEエラーが出ているので確認して修正してください。
```
19-Oct-2020 18:11:11 Europe/Berlin] PHP Notice:  Undefined variable: message in /Applications/MAMP/htdocs/mizutaniyu_php/php-rt-iine/index.php on line 93
[19-Oct-2020 18:11:11 Europe/Berlin] PHP Notice:  Undefined index: res in /Applications/MAMP/htdocs/mizutaniyu_php/php-rt-iine/index.php on line 94
```

U-11 replied:
$_POST['retweet']及び$_REQUEST['res']のNULL判定を行いNOTICEエラーを解消しました。

gacha1987 commented on php-rt-iine/like_action.php:
```diff
@@ -0,0 +1,62 @@
+    
+    header('Location:'.$url);
+    exit();
+  }
```
今の実装ですと、仮にlike_memberの値に数値以外が入っていた場合、exitも呼ばれず処理が最後まで進み、ただ真っ白な画面になってしまいます。
エラーメッセージ表示でもTOPにリダイレクトさせるでもいいのでバリデーションエラーだった場合の考慮を追加してみましょう！

判定式が多いとルート漏れが発生する可能性があるので他の箇所も問題ないか確認してみてください。

U-11 replied:
バリデーションエラーだった場合にTOPにリダイレクトする実装を行いました。

gacha1987 replied:
非常に惜しいです。
`delete_like`と`like_add `がパラメータに指定されなかった場合にまだ真っ白になります。
リツイートも同様です。

gacha1987 commented on php-rt-iine/rt_action.php:
```diff
@@ -0,0 +1,100 @@
+  }
+}else{
+// コメントなしリツイートのDB登録
+  if(is_numeric($_POST['rt_post_id']) && is_numeric($_POST['post_member_id'])){
```
リツイートを消す時にNOTICEエラーが出るので修正してください。
```
[19-Oct-2020 18:33:03 Europe/Berlin] PHP Notice:  Undefined index: rt_post_id in /Applications/MAMP/htdocs/mizutaniyu_php/php-rt-iine/rt_action.php on line 58
```

U-11 replied:
NULL判定を行いNOTICEエラーを解消しました。

gacha1987 commented on php-rt-iine/style.css:
```diff
@@ -0,0 +1,227 @@
+
+
+body {
+	background: url(images/bg_main.png) top center repeat-y #eeeeee;
```
ブラウザの開発者ツールで見るとimages配下がないため404になっているのでこちらもコミット対象に含めてください。
![スクリーンショット 0002-10-20 1 36 17](https://user-images.githubusercontent.com/8597103/96480053-ac3b3600-1274-11eb-8b21-cd530eb85ef7.png)

U-11 replied:
images配下ファイルをGit管理対象に追加

gacha1987 reviewed (commented):
コメントしました。徐々にコードが見やすくなってきていていい感じです！
Noticeエラーが所々で発生しています。リツイートを削除した時のみ発生する等特定条件で発生する場合もあるので色々試しながらチェックしてみてください。

Commit:
バリデーションエラーだった場合TOPにリダイレクト

Commit:
三項演算子の書き方訂正

Commit:
Noticeエラーの修正、バリデーションエラーの場合TOPにリダイレクト

Commit:
NOTICEエラーの修正

Commit:
CSSのimages追加及び文末の不要な書き込みの削除

gacha1987 commented:
こちらの対応が不十分のため、追加コメントしましたので確認して修正してください。
https://github.com/U-11/mizutaniyu_php/pull/1#discussion_r509958244

Commit:
$#_POSTパラメータに値が指定がなされなかった際にTOPリダイレクトするよう実装

Commit:
$_POSTパラメータに値が指定がなされなかった際にTOPリダイレクトするよう実装

Commit:
Merge branch 'feature/rt-iine' of https://github.com/U-11/mizutaniyu_php into feature/rt-iine

gacha1987 commented on php-rt-iine/like_action.php:
```diff
@@ -44,6 +44,9 @@
   }
+}else{
+  header('Location:index.php');
+  exit();
```
ここをelseにしてexitしてしまうと、いいね取り消しのルートに入れないのでいいねが消せない状態になっていますので修正してください。
過去正常に動いた箇所でも修正を入れることでバグが起きることがあるので修正を入れたら挙動の確認は一通り行いましょう。(こういうのをデグレードと言います)

U-11 replied:
デグレードを行いバグを修正しました。

gacha1987 replied:
修正確認しました。
すみません、、自分の書き方が悪かったです 🙇 デグレードとは修正起因で新たなバグが発生することを指します。
http://e-words.jp/w/%E3%83%87%E3%82%B0%E3%83%AC%E3%83%BC%E3%83%89.html#:~:text=%E3%83%87%E3%82%B0%E3%83%AC%E3%83%BC%E3%83%89%E3%81%A8%E3%81%AF%E3%80%81%E6%96%B0%E3%81%97%E3%81%84%E3%83%90%E3%83%BC%E3%82%B8%E3%83%A7%E3%83%B3,%E3%82%92%E6%8C%87%E3%81%99%E5%A0%B4%E5%90%88%E3%82%82%E3%81%82%E3%82%8B%E3%80%82

gacha1987 reviewed (commented):
コメントしました。

Commit:
デグレードを行いバグを修正しました。

gacha1987 commented on php-rt-iine/like_action.php:
```diff
@@ -0,0 +1,71 @@
+  $_SESSION['time']=time();
+
+  $members=$db->prepare('SELECT * FROM members WHERE id=?');
+  $members->execute(array($_SESSION['id']));
```
**修正不要です**
配列の初期化は最近は `array()`ではなく`[]`で行うのがメジャーです。
array($_SESSION['id']) → [$_SESSION['id']]

gacha1987 reviewed (approved):
LGTMです！
コードの記述スタイルが統一されていたので、今回特に指摘はしていませんが、
PHPを使う現場では主にPSRという標準規約を採用しているところが多いので、
次に実装してみるときはそれに基づいてコーディングしてみるとより実践的になると思うので気が向いたら試してみてください！
※PSRは種類が多いのでいきなり調べると混乱するかもですが、ひとまずPSR-12に準拠できれば大丈夫です。
https://www.ritolab.com/entry/208
